File: src/waypalm-site/src/main/java/waypalm/site/web/mvc/PrincipalProviderOpenID.java
package waypalm.site.web.mvc;

import com.iserv2.commons.lang.IservHashUtil;
import com.iserv2.commons.lang.Predicate;
import com.iserv2.commons.lang.collections.IservCollections;
import org.springframework.security.core.userdetails.AuthenticationUserDetailsService;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.openid.OpenIDAttribute;
import org.springframework.security.openid.OpenIDAuthenticationToken;
import org.springframework.social.connect.ConnectionKey;
import waypalm.common.web.model.SocialProfile;
import waypalm.domain.entity.Profile;
import waypalm.domain.entity.SocialLink;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.List;
import java.util.Locale;

public class PrincipalProviderOpenID extends PrincipalProvider implements AuthenticationUserDetailsService<OpenIDAuthenticationToken> {
    @Override
    public UserDetails loadUserDetails(OpenIDAuthenticationToken token) throws UsernameNotFoundException {
        return createPrincipal(loadUser(token));
    }

    private String createOpenIDProvider(URL url) {
        return url.getHost().toLowerCase(Locale.ENGLISH);
    }

    private String createOpenIDUserName(URL url) {
        return IservHashUtil.md5(url.toString());
    }

    private Profile loadUser(OpenIDAuthenticationToken token) {
        URL url;
        try {
            url = new URL(token.getIdentityUrl());
        } catch (MalformedURLException e) {
            throw new IllegalArgumentException("wrong identity url");
        }

        String provider = createOpenIDProvider(url);
        String userName = createOpenIDUserName(url);
        String email = takeFirstAttribute(token.getAttributes(), "email");

        if (email == null) {
            email = takeFirstAttribute(token.getAttributes(), "axEmail");
        }

        SocialLink socialLink = userRepository.findSocialConnection(provider, userName, email);
        if (socialLink == null) {
            ConnectionKey socialKey = new ConnectionKey(provider, userName);
            SocialProfile socialProfile = new SocialProfile();
            socialProfile.setUrl(token.getIdentityUrl());
            socialProfile.setEmail(email);
            socialLink = userService.createSocialConnection(getLoggedUser(), socialKey, socialProfile);
        }
        return socialLink.getProfile();
    }

    private String takeFirstAttribute(List<OpenIDAttribute> attributes, final String name) {
        OpenIDAttribute attr = IservCollections.findFirst(attributes, new Predicate<OpenIDAttribute>() {
            @Override
            public boolean test(OpenIDAttribute item) {
                return name.equals(item.getName());
            }
        });
        return attr == null || attr.getCount() < 1 ? null : attr.getValues().get(0);
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/web/mvc/PrincipalProviderOpenID.kt
package waypalm.site.web.mvc

import javax.inject.Qualifier
import java.lang.annotation.RetentionPolicy
import java.lang.annotation.Retention
import org.springframework.security.core.userdetails.AuthenticationUserDetailsService
import org.springframework.security.openid.OpenIDAuthenticationToken
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.transaction.annotation.Transactional
import waypalm.domain.entity.Profile
import java.net.URL
import java.net.MalformedURLException
import waypalm.common.web.model.SocialProfile
import org.springframework.social.connect.ConnectionKey
import org.springframework.security.openid.OpenIDAttribute
import waypalm.common.util.ifNull
import java.util.Locale
import com.iserv2.commons.lang.IservHashUtil
import javax.inject.Inject
import waypalm.site.services.UserAccess
import waypalm.domain.dao.UserRepository
import waypalm.site.services.UserService
import javax.inject.Named

Named OpenID
public class PrincipalProviderOpenID [Inject](
        var userService: UserService,
        var userRepository: UserRepository,
        var auth: UserAccess
)
: PrincipalProvider()
, AuthenticationUserDetailsService<OpenIDAuthenticationToken>
{
    Transactional
    public override fun loadUserDetails(token: OpenIDAuthenticationToken?): UserDetails? {
        return createPrincipal(loadUser(token!!))
    }

    fun loadUser(token: OpenIDAuthenticationToken): Profile {
        var url: URL;
        try {
            url = URL(token.getIdentityUrl());
        } catch (e: MalformedURLException) {
            throw IllegalArgumentException("wrong identity url");
        }

        val provider = createOpenIDProvider(url);
        val userName = createOpenIDUserName(url);
        val email = takeFirstAttribute(token.getAttributes(), "email").ifNull(takeFirstAttribute(token.getAttributes(), "axEmail").ifNull(""));

        val socialLink = userRepository.findSocialConnection(provider, userName, email);
        if (socialLink == null) {
            return userService.createSocialConnection(auth.loadCurrentProfile()
                    , ConnectionKey(provider, userName)
                    , SocialProfile(url = token.getIdentityUrl()!!, email = email)
            )!!.profile;
        }
        return socialLink.profile;
    }

    fun takeFirstAttribute(attributes: List<OpenIDAttribute>?, name: String): String? {
        var attr = attributes!!
                .find{ name.equals(it.getName()) && it.getCount() > 1 }
        return attr?.getValues()!!.get(0)
    }

    fun createOpenIDProvider(url: URL): String {
        return url.getHost()!!.toLowerCase(Locale.ENGLISH);
    }

    fun createOpenIDUserName(url: URL): String {
        return IservHashUtil.md5(url.toString())!!;
    }
}

Retention(RetentionPolicy.RUNTIME)
Qualifier
public annotation class OpenID
